### PR TITLE
AP_DDS: Support sensor_msgs/BatteryState message

### LINKED ISSUE #23360: AP_DDS: Add support for sensor_msgs/BatteryState
## Feature request
Add support in DDS for publishing Battery Voltage from Ardupilot. Similar to NavSatFix, only use the first battery. Handling multiple GPS's or multiple batteries can be solved later. 

**Describe the solution you'd like**

Publish battery information at a low rate (1Hz). 

**Describe alternatives you've considered**

This is an alternative to mavlink

**Platform**
[ X ] All
[  ] AntennaTracker
[  ] Copter
[  ] Plane
[  ] Rover
[  ] Submarine

### PR DESCRIPTION
**Pull request by pedro-fuoco**
Resolves #23360 
This is just a draft, I still need to integrate the battery data from AP_BattMonitor
For now it looks like this:
Screenshot from 2023-03-30 13-03-28

### CODE REVIEW

**Comment by pedro-fuoco** [comment]
Screenshot from 2023-03-30 17-55-23

File: libraries/AP_DDS/Idl/BatteryState.idl
Comment (Ryanf55) [inline libraries/AP_DDS/Idl/BatteryState.idl:1]: If you update your MicroXRCEDDSGen to `develop`, I think you can revert these changes.

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 1]
```diff
@@ -120,6 +122,76 @@ void AP_DDS_Client::update_topic(sensor_msgs_msg_NavSatFix& msg, const uint8_t i
...
+    msg.power_supply_technology = 0; //POWER_SUPPLY_TECHNOLOGY_UNKNOWN
+
+
+    if (battery.has_cell_voltages(instance))
+    {
+        const uint16_t* intVoltages = battery.get_cell_voltages(instance).cells;
+        float floatVoltages[sizeof(intVoltages)];
+        for (int i=0; i < (int)sizeof(intVoltages); i++) 
+        {
+            floatVoltages[i] = (float)intVoltages[i];
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:190]: prefer a `static_cast<float>`

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 2]
```diff
@@ -120,6 +122,76 @@ void AP_DDS_Client::update_topic(sensor_msgs_msg_NavSatFix& msg, const uint8_t i
...
+    msg.power_supply_health = (battery.overpower_detected(instance)) ? 4 : 1; //POWER_SUPPLY_HEALTH_OVERVOLTAGE or POWER_SUPPLY_HEALTH_GOOD
+
+    msg.power_supply_technology = 0; //POWER_SUPPLY_TECHNOLOGY_UNKNOWN
+
+
+    if (battery.has_cell_voltages(instance))
+    {
+        const uint16_t* intVoltages = battery.get_cell_voltages(instance).cells;
+        float floatVoltages[sizeof(intVoltages)];
+        for (int i=0; i < (int)sizeof(intVoltages); i++)
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:188]: Want to use the array size macro `ARRAY_SIZE`?

File: libraries/AP_DDS/AP_DDS_Client.cpp (hunk 1 above)
Comment (pedro-fuoco) [reply libraries/AP_DDS/AP_DDS_Client.cpp:190]: Just realized I might not need to cast at all, as std::copy will already cast implicitly. So something like this would be enough:
```
const uint16_t* cellVoltages = battery.get_cell_voltages(instance).cells;
std::copy(std::begin(cellVoltages),std::end(cellVoltages),std::begin(msg.cell_voltage));
```

File: libraries/AP_DDS/AP_DDS_Client.cpp (hunk 2 above)
Comment (pedro-fuoco) [reply libraries/AP_DDS/AP_DDS_Client.cpp:188]: Looking into AP_BattMonitor.h I found that the cells struct is defined like this:
```
// cell voltages in millivolts
struct cells {
    uint16_t cells[AP_BATT_MONITOR_CELLS_MAX];
};
```
So getting the size of the array is unnecessary, as it is always AP_BATT_MONITOR_CELLS_MAX

File: libraries/AP_DDS/Idl/BatteryState.idl
Comment (pedro-fuoco) [reply libraries/AP_DDS/Idl/BatteryState.idl:1]: I am already at develop, but I couldn't get neither serial_number nor cell_voltage to work properly and I suspect it has something to do with the DDS.
I've tried to replicating what you did at NavSatFix's msg.header.frame_id and msg.position_covariance, but with no success. Can you give me a hand ?

File: libraries/AP_DDS/AP_DDS_Client.cpp (hunk 1 above)
Comment (pedro-fuoco) [reply libraries/AP_DDS/AP_DDS_Client.cpp:190]: how about using the implicit transformation in std::copy ? 
Like so:
```
if (battery.has_cell_voltages(instance))
{
   const uint16_t* cellVoltages = battery.get_cell_voltages(instance).cells;
   std::copy(cellVoltages, cellVoltages + AP_BATT_MONITOR_CELLS_MAX, msg.cell_voltage);
} 
```
Also solves the array size issue bellow

Comment (Ryanf55) [reply libraries/AP_DDS/AP_DDS_Client.cpp:190]: Yes, that looks great.

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 3]
```diff
@@ -120,6 +122,70 @@ void AP_DDS_Client::update_topic(sensor_msgs_msg_NavSatFix& msg, const uint8_t i
...
 }
 
+void AP_DDS_Client::update_topic(sensor_msgs_msg_BatteryState& msg, const uint8_t instance)
+{
+    if (instance >= AP_BATT_MONITOR_MAX_INSTANCES) {
+        return;
+    }
+
+    update_topic(msg.header.stamp);
+    auto &battery = AP::battery();
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:132]: Before you get an instance of the battery, do you want to lock the HAL semaphore? 

Similar to the above `WITH_SEMAPHORE(gps.get_semaphore());`

This ensures the query is done all at the same timestamp, from what I understand.

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 4]
```diff
@@ -120,6 +122,70 @@ void AP_DDS_Client::update_topic(sensor_msgs_msg_NavSatFix& msg, const uint8_t i
...
+        msg.percentage = NAN;
+        msg.charge = NAN;
+    }
+    
+    msg.design_capacity = NAN;
+ 
+    if (battery.current_amps(current, instance))
+    {
+        if (percentage == 100) {msg.power_supply_status = 4;} //POWER_SUPPLY_STATUS_FULL
+        else if (-1*current > 0.0) {msg.power_supply_status = 1;} //POWER_SUPPLY_STATUS_CHARGING
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:170]: Why not flip logic, removing an extra multiplication?
```suggestion
        else if (current > 0.0) {msg.power_supply_status = 1;} //POWER_SUPPLY_STATUS_CHARGING
```

**Review by Ryanf55 — COMMENTED** [review]
Getting close. Can you show your test results (screenshot is fine) here?

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 5]
```diff
@@ -120,6 +122,70 @@ void AP_DDS_Client::update_topic(sensor_msgs_msg_NavSatFix& msg, const uint8_t i
...
+    msg.voltage = battery.voltage(instance);
+    
+    float temperature; 
+    msg.temperature = (battery.get_temperature(temperature, instance)) ? temperature : NAN;
+    
+    float current;
+    msg.current = (battery.current_amps(current, instance)) ? -1 * current : NAN;
+    
+    float capacity = (float)battery.pack_capacity_mah(instance)/1000.0;
+    msg.capacity = capacity;
```
[the remark this answers is not in the document]
Comment (pedro-fuoco) [reply libraries/AP_DDS/AP_DDS_Client.cpp:151]: the `capacity` variable is used later:
`msg.charge = (percentage) * (capacity);`

File: libraries/AP_DDS/AP_DDS_Client.cpp (hunk 4 above)
Comment (pedro-fuoco) [reply libraries/AP_DDS/AP_DDS_Client.cpp:170]: fair point, I thought it would be easier for people to read it at first, but I see how it isn't

File: libraries/AP_DDS/AP_DDS_Client.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ryanf55) [reply libraries/AP_DDS/AP_DDS_Client.cpp:151]: Can you make it `const` at least?

[the remark this answers is not in the document]
Comment (pedro-fuoco) [reply libraries/AP_DDS/AP_DDS_Client.cpp]: Yep, fixed it up on the first commit

File: libraries/AP_DDS/Idl/BatteryState.idl
Comment (Ryanf55) [reply libraries/AP_DDS/Idl/BatteryState.idl:1]: For cell_voltage, set  attribute `cell_voltage_size` equal to the number of cells the battery has.
For serial_number, you should be able use `get_serial_number` and then `sprintf` to write it into the char*.

**Comment by Ryanf55** [comment]
I've tested this on Hardware with CubeOrange+. Voltage works fine. 

Here is the parameter I needed to enable
`BATT_MONITOR     4.0`

I am unable to set `BATT_CAPACITY` on ArduPlane, the param does not show up. Same issue with pretty much all the BATT_* params. 

Anyways, here's the data output. I confirmed the constants output in this case are sane. 
```
---
header:
  stamp:
    sec: 1210
    nanosec: 496124000
  frame_id: ''
voltage: 11.610426902770996
temperature: .nan
current: -0.025501174852252007
charge: 3.2669999599456787
capacity: 3.299999952316284
design_capacity: .nan
percentage: 99.0
power_supply_status: 2
power_supply_health: 1
power_supply_technology: 0
present: true
cell_voltage: []
cell_temperature: []
location: ''
serial_number: ''
---
```

Until we have really improved the simulation ability, I don't have an exact way to test the rest of the PR, however from visual inspection, all of it looks good. 

One question. Why is `design_capacity` not filled in? This should be set from the param `BATT_CAPACITY` , but that's not even on my vehicle, or in the source code for anything but rover, sim, and some logs. 

Approved!

**Comment by pedro-fuoco** [comment]
Thank you so much for the hardware tests @Ryanf55 ! It's awesome to see it working. 
I've squashed the other changes to have this merge-ready

Regarding your question about `msg.design_capacity`:
Currently, `msg.capacity` is filled in with `AP_BattMonitor`'s `pack_capacity_mah`
```
/// pack_capacity_mah - returns the capacity of the battery pack in mAh when the pack is full
int32_t AP_BattMonitor::pack_capacity_mah(uint8_t instance) const
{
    if (instance < AP_BATT_MONITOR_MAX_INSTANCES) {
        return _params[instance]._pack_capacity;
    } else {
        return 0;
    }
}
```
Notice how `_params[instance]._pack_capacity` is already the value from the `BATT_CAPACITY` ardupilot param
```
AP_GROUPINFO("CAPACITY", 7, AP_BattMonitor_Params, _pack_capacity, 3300),
```

I could fill `msg.design_capacity` the same way to get the value from `BATT_CAPACITY`, but then `msg.capacity` and `msg.design_capacity` would be the same.

**Comment by Ryanf55** [comment]
> pack_capacity_mah

Based on the comments, I think you should instead populate design_capacity, since that's what written on the sticker on the battery and what a user usually inputs as the param. 

I don't think you can measure `msg.capacity` in ArduPilot, since ArduPilot isn't a battery charger, so perhaps leave it blank. That would be the number displayed on the charger in the case you had a dead battery and charged it full. This could be smaller than `design_capacity` if the battery is old, or the manufacturer is overly optimistic in their specs. Thus, perhaps it should be NAN. 

```
#float32 capacity # Capacity in Ah (last full capacity) (If unmeasured NaN)
#float32 design_capacity # Capacity in Ah (design capacity) (If unmeasured NaN)
```

**Comment by pedro-fuoco** [comment]
Fixed up msg.design_capacity and merged the PR to work with the static transformations
Screenshot from 2023-04-13 11-41-16

**Comment by pedro-fuoco** [comment]
Fixed some CI problems.
I think it's finally ready to merge on CI pass

File: libraries/AP_DDS/AP_DDS_Topic_Table.h [hunk 6]
```diff
@@ -2,6 +2,7 @@
 #include "sensor_msgs/msg/NavSatFix.h"
 #include "tf2_msgs/msg/TFMessage.h"
 
+#include "generated/BatteryState.h"
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Topic_Table.h:5]: Wrong header guard with the new file reorg

Comment (pedro-fuoco) [reply libraries/AP_DDS/AP_DDS_Topic_Table.h]: yep, was fixing it rn. Tested it this time, its working. Sorry about that

**Review by Ryanf55 — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks!
